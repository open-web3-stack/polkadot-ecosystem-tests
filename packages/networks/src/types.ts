import { SetupOption } from '@acala-network/chopsticks-testing'

type ChainConfigRelaychain = {
  isRelayChain: true
  paraId?: undefined
}

type ChainConfigParachain = {
  isRelayChain?: false
  paraId: number
}

type ChainConfigBase = {
  name: string
  endpoint: string | string[]
  isRelayChain?: boolean
} & (ChainConfigRelaychain | ChainConfigParachain)

export type ChainConfig<
  TCustom extends Record<string, unknown> | undefined,
  TInitStorages extends Record<string, Record<string, any>> | undefined,
> = ChainConfigBase &
  SetupOption & {
    custom: TCustom
    initStorages?: TInitStorages
  }

export type Chain<
  TCustom extends Record<string, unknown> | undefined = Record<string, unknown> | undefined,
  TInitStorages extends Record<string, Record<string, any>> | undefined =
    | Record<string, Record<string, any>>
    | undefined,
  TExtended extends Record<string, unknown> | undefined = Record<string, unknown> | undefined,
> = ChainConfig<TCustom, TInitStorages> &
  TExtended & {
    extend: <config extends Record<string, unknown>>(
      fn: (chain: Chain<TCustom, TInitStorages, TExtended>) => config,
    ) => Chain<TCustom, TInitStorages, TExtended & config>
  }
