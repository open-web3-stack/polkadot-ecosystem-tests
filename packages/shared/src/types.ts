import type { Blockchain, StorageValues } from '@acala-network/chopsticks'
import type { Chain } from '@e2e-test/networks'
import type { ApiPromise, WsProvider } from '@polkadot/api'

import { match } from 'ts-pattern'
import { afterAll, beforeAll, describe, test } from 'vitest'

export type Prettify<T> = {
  [K in keyof T]: T[K]
} & {}

/**
 * Type of client object produced by `setupNetworks` in E2E test suites.
 */
export type Client<
  TCustom extends Record<string, unknown> | undefined,
  TInitStorages extends Record<string, Record<string, any>> | undefined,
> = {
  config: Chain<TCustom, TInitStorages>
  url: string
  chain: Blockchain
  ws: WsProvider
  api: ApiPromise
  dev: {
    newBlock: (param?: Partial<any>) => Promise<string>
    setStorage: (values: StorageValues, blockHash?: string) => Promise<any>
    timeTravel: (date: string | number) => Promise<number>
    setHead: (hashOrNumber: string | number) => Promise<any>
  }
  teardown(): Promise<void>
  pause(): Promise<unknown>
}

type TestNode = {
  kind: 'test'
  /**
   * The label used to identify the test node when `vitest.test` is called.
   */
  label: string
  /**
   * A function returning a promise (actual test body).
   * This is passed into `vitest.test(...)` during registration.
   */
  testFn: () => Promise<void>
  flags?: { only?: boolean; skip?: boolean; timeout?: number }
  meta?: Record<string, any>
}

type DescribeNode = {
  kind: 'describe'
  label: string
  children: Node[]
  beforeAll?: () => Promise<void>
  afterAll?: () => Promise<void>
  flags?: { only?: boolean; skip?: boolean }
  meta?: Record<string, any>
}

/**
 * A test tree, used to represent an E2E test suite.
 */
type Node = TestNode | DescribeNode

export function runNode(node: Node) {
  match(node)
    .with({ kind: 'test' }, (testNode) => {
      const t = testNode.flags?.only ? test.only : testNode.flags?.skip ? test.skip : test

      // Recall this is `test` from `vitest`
      t(testNode.label, { timeout: testNode.flags?.timeout }, testNode.testFn)
    })
    .with({ kind: 'describe' }, (describeNode) => {
      const d = describeNode.flags?.only ? describe.only : describeNode.flags?.skip ? describe.skip : describe

      // Recall this is `describe` from `vitest`
      d(describeNode.label, () => {
        if (describeNode.beforeAll) beforeAll(describeNode.beforeAll)
        if (describeNode.afterAll) afterAll(describeNode.afterAll)
        describeNode.children.forEach(runNode)
      })
    })
    .exhaustive()
}
