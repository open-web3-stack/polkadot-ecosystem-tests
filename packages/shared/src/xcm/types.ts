import type { ApiPromise } from '@polkadot/api'
import type { SubmittableExtrinsic } from '@polkadot/api/types'

export type Tx = ({ api }: { api: ApiPromise }, acc: any) => SubmittableExtrinsic<'promise'>
export type GetBalance = ({ api }: { api: ApiPromise }, address: string) => Promise<any>
