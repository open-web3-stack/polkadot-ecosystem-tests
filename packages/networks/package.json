{
	"name": "@e2e-test/networks",
	"type": "module",
	"dependencies": {
		"@acala-network/chopsticks-testing": "^0.13.1",
		"@e2e-test/shared": "workspace:*"
	},
	"main": "./dist/cjs/index.js",
	"module": "./dist/esm/index.js",
	"files": [
		"*",
		"!tsconfig.json",
		"!**/*.test.ts"
	],
	"exports": {
		".": {
			"require": "./dist/cjs/index.js",
			"import": "./dist/esm/index.js",
			"default": "./dist/esm/index.js"
		},
		"./chains": {
			"require": "./dist/cjs/chains/index.js",
			"import": "./dist/esm/chains/index.js",
			"default": "./dist/esm/chains/index.js"
		},
		"./*": {
			"require": "./dist/cjs/*.js",
			"import": "./dist/esm/*.js",
			"default": "./dist/esm/*.js"
		}
	}
}
