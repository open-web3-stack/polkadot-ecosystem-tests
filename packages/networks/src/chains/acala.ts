import { defaultAccount } from '@e2e-test/shared'

import { defineChain } from '../defineChain.js'

const custom = {
  acala: {
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
    stableToken: 'AUSD',

    paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
    dot: { Token: 'DOT' },
    ldot: { Token: 'LDOT' },
    dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
    wbtc: { ForeignAsset: 5 },
    ausd: { Token: 'AUSD' },
    aca: { Token: 'ACA' },
    lcdot: { LiquidCrowdloan: 13 },
  },
  karura: {
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
    stableToken: 'KUSD',

    paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
    ksm: { Token: 'KSM' },
    lksm: { Token: 'LKSM' },
    usdt: { ForeignAsset: 7 },
    rmrk: { ForeignAsset: 0 },
    dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
    ausd: { Token: 'KUSD' },
    kar: { Token: 'KAR' },
  },
}

const getInitStorages = (config: typeof custom.acala | typeof custom.karura) => ({
  System: {
    account: [[[defaultAccount.alice.address], { providers: 4, data: { free: 10 * 1e12 } }]],
  },
  Tokens: {
    accounts: [
      [[defaultAccount.alice.address, { Token: config.relayToken }], { free: 10 * 1e12 }],
      [[defaultAccount.alice.address, { Token: config.relayLiquidToken }], { free: 100 * 1e12 }],
      [[defaultAccount.alice.address, { Token: config.stableToken }], { free: 1000 * 1e12 }],
    ],
  },
  Sudo: {
    key: defaultAccount.alice.address,
  },
  EvmAccounts: {
    accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], defaultAccount.alice.address]],
    evmAddresses: [[[defaultAccount.alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
  },
  Homa: {
    // avoid impact test outcome
    $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
    // so that bump era won't trigger unbond
    relayChainCurrentEra: 100,
  },
  PolkadotXcm: {
    // avoid sending xcm version change notifications to makes things faster
    $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
  },
})

export const acala = defineChain({
  name: 'acala',
  endpoint: 'wss://acala-rpc.aca-api.network',
  paraId: 2000,
  custom: custom.acala,
  initStorages: getInitStorages(custom.acala),
})

export const karura = defineChain({
  name: 'karura',
  endpoint: 'wss://karura-rpc.aca-api.network',
  paraId: 2000,
  custom: custom.karura,
  initStorages: getInitStorages(custom.karura),
})
