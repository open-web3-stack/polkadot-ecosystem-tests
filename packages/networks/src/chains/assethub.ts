import { defaultAccounts } from '../defaultAccounts.js'
import { defineChain } from '../defineChain.js'

const custom = {
  assetHubPolkadot: {
    dot: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
  assetHubKusama: {
    ksm: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
}

const getInitStorages = (config: typeof custom.assetHubPolkadot | typeof custom.assetHubKusama) => ({
  System: {
    account: [[[defaultAccounts.alice.address], { providers: 1, data: { free: 1000e10 } }]],
  },
  Assets: {
    account: [
      [[config.usdtIndex, defaultAccounts.alice.address], { balance: 1000e6 }], // USDT
    ],
  },
})

export const assetHubPolkadot = defineChain({
  name: 'assetHubPolkadot',
  endpoint: 'wss://rpc-asset-hub-polkadot.luckyfriday.io',
  paraId: 1000,
  custom: custom.assetHubPolkadot,
  initStorages: getInitStorages(custom.assetHubPolkadot),
})

export const assetHubKusama = defineChain({
  name: 'assetHubKusama',
  endpoint: 'wss://rpc-asset-hub-kusama.luckyfriday.io',
  paraId: 1000,
  custom: custom.assetHubKusama,
  initStorages: getInitStorages(custom.assetHubKusama),
})
