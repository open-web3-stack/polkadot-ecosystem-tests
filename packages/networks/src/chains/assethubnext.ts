import { defaultAccounts, defaultAccountsSr25519 } from '../defaultAccounts.js'
import { defineChain } from '../defineChain.js'

const custom = {
  assetHubNextWestend: {
    wnd: { Concrete: { parents: 1, interior: 'Here' } },
    usdt: { Concrete: { parents: 0, interior: { X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
    usdtIndex: 1984,
  },
}

const getInitStorages = (config: typeof custom.assetHubNextWestend) => ({
  System: {
    account: [
      [[defaultAccounts.alice.address], { providers: 1, data: { free: 1000e10 } }],
      [[defaultAccountsSr25519.alice.address], { providers: 1, data: { free: 1000e10 } }],
    ],
  },
  Assets: {
    account: [
      [[config.usdtIndex, defaultAccounts.alice.address], { balance: 1000e6 }], // USDT
    ],
  },
})

export const assetHubNextWestend = defineChain({
  name: 'assetHubNextWestend',
  endpoint: 'wss://westend-asset-hub-next-rpc.parity-chains-scw.parity.io',
  paraId: 1100,
  custom: custom.assetHubNextWestend,
  initStorages: getInitStorages(custom.assetHubNextWestend),
})
