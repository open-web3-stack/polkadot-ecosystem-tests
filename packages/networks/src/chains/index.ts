export * from './acala.js'
export * from './assethub.js'
export * from './astar.js'
export * from './hydraDX.js'
export * from './moonbeam.js'
export * from './polkadot.js'
