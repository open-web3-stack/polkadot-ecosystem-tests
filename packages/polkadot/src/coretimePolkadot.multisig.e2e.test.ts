import { coretimePolkadot } from '@e2e-test/networks/chains'

import { baseMultisigE2Etests } from '@e2e-test/shared'
import { registerTestTree } from '@e2e-test/shared/types'

registerTestTree(
  baseMultisigE2Etests(coretimePolkadot, { testSuiteName: 'CoretimePolkadot Multisig', addressEncoding: 0 }),
)
