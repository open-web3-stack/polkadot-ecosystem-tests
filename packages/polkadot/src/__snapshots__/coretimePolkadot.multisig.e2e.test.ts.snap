// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoretimePolkadot Multisig > approval with max weight too low fails > events when Bob executes multisig operation with low weight 1`] = `[]`;

exports[`CoretimePolkadot Multisig > approval with sender in signatories fails > events when approval with sender in signatories fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x06000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2400000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > approval with signatories out of order fails > events when approval with signatories out of order fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x05000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2400000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > approval with wrong timepoint fails > events when Bob executes multisig operation with wrong block number 1`] = `[]`;

exports[`CoretimePolkadot Multisig > approval with wrong timepoint fails > events when Bob executes multisig operation with wrong extrinsic index 1`] = `[]`;

exports[`CoretimePolkadot Multisig > approval without timepoint fails > events when Bob executes multisig operation with \`approveAsMulti\` 1`] = `[]`;

exports[`CoretimePolkadot Multisig > basic 2-of-3 multisig creation and execution > events when Alice approves multisig call 1`] = `
[
  {
    "data": {
      "approving": "14E5nqKAp3oAJcmzgZhUD2RcptBeUBScxKHgJKU4HPNcKVf3",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
      "result": "Ok",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigExecuted",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > basic 2-of-3 multisig creation and execution > events when Alice creates multisig 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > beginning multisig approval with \`approveAsMulti\` works > events when Alice starts multisig with \`approveAsMulti\` 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > beginning multisig approval with \`approveAsMulti\` works > events when Bob executes with \`asMulti\` 1`] = `
[
  {
    "data": {
      "approving": "14E5nqKAp3oAJcmzgZhUD2RcptBeUBScxKHgJKU4HPNcKVf3",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
      "result": "Ok",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigExecuted",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > cancellation with sender in signatories fails > events for cancellation with sender in signatories fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x06000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > cancelling a non-existent multisig operation fails > events when cancelling multisig with bogus call hash fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x07000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > cancelling a non-existent multisig operation fails > events when cancelling multisig with wrong signatories fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x07000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > creating a multisig with threshold < 2 fails > events when creating multisig with threshold < 2 fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x00000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2500000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > execution with sender in signatories fails > events when creation with sender in signatories fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x06000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2500000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > final approval with \`approveAsMulti\` does not lead to execution > events when Bob makes final approval with approveAsMulti 1`] = `
[
  {
    "data": {
      "approving": "14E5nqKAp3oAJcmzgZhUD2RcptBeUBScxKHgJKU4HPNcKVf3",
      "callHash": "(hash)",
      "multisig": "13yhdvhAzML3u5MENoG8WCT9bTVT5H3C6pjdMpy4PZNisgbS",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigApproval",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > first call with unexpected timepoint fails > events when Alice starts multisig operation with \`approveAsMulti\` 1`] = `[]`;

exports[`CoretimePolkadot Multisig > multisig cancellation with remaining signatories out of order fails > events when Alice creates multisig for cancel ordering test 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig cancellation with remaining signatories out of order fails > events when multisig cancellation with signatories out of order fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x05000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig cancellation with threshold < 2 fails > events when multisig cancellation with threshold < 2 fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x00000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig cancellation works > events when Alice cancels multisig 1`] = `
[
  {
    "data": {
      "callHash": "(hash)",
      "cancelling": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigCancelled",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig cancellation works > events when Alice creates multisig for cancellation 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig creation with too many signatories fails > events when creating multisig with too many signatories fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x04000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2500000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig execution with remaining signatories out of order fails > events when Alice creates multisig for ordering test 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > multisig execution with remaining signatories out of order fails > events when multisig approval with signatories out of order fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x05000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2500000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > non-depositor tries to cancel multisig fails > events when non-depositor tries to cancel multisig fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x08000000",
          "index": "(rounded 40)",
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 20000)",
          "refTime": "(rounded 1000000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > repeated approval with \`approveAsMulti\` fails > events when repeated approval with approveAsMulti fails 1`] = `
[
  {
    "data": {
      "dispatchError": {
        "Module": {
          "error": "0x01000000",
          "index": 41,
        },
      },
      "dispatchInfo": {
        "class": "Normal",
        "paysFee": "Yes",
        "weight": {
          "proofSize": "(rounded 1000000)",
          "refTime": "(rounded 2400000000)",
        },
      },
    },
    "method": "ExtrinsicFailed",
    "section": "system",
  },
]
`;

exports[`CoretimePolkadot Multisig > second approval (with \`approveAsMulti\`) in 2-of-3 multisig is successful and does not lead to execution > events when Alice creates multisig for approveAsMulti test 1`] = `
[
  {
    "data": {
      "approving": "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
    },
    "method": "NewMultisig",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > second approval (with \`approveAsMulti\`) in 2-of-3 multisig is successful and does not lead to execution > events when Bob approves with approveAsMulti 1`] = `
[
  {
    "data": {
      "approving": "14E5nqKAp3oAJcmzgZhUD2RcptBeUBScxKHgJKU4HPNcKVf3",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigApproval",
    "section": "multisig",
  },
]
`;

exports[`CoretimePolkadot Multisig > second approval (with \`approveAsMulti\`) in 2-of-3 multisig is successful and does not lead to execution > events when Charlie provides final approval 1`] = `
[
  {
    "data": {
      "approving": "14Gjs1TD93gnwEBfDMHoCgsuf1s2TVKUP6Z1qKmAZnZ8cW5q",
      "callHash": "(hash)",
      "multisig": "12fqSn9qVLJL4NY7Uua7bexEAVr9oCpD3e5xmdpNjtQszzBt",
      "result": "Ok",
      "timepoint": {
        "height": "(redacted)",
        "index": 2,
      },
    },
    "method": "MultisigExecuted",
    "section": "multisig",
  },
]
`;
